### PR TITLE
Tuning with sweep(Completed)

### PR DESCRIPTION
**Pull request by Wxy-23**
Still in the debugging phase of implementing wandb.

### CODE REVIEW

**Comment by Wxy-23** [comment]
**Hyperparameter Tuning Implementation Using W&B Sweeps**

Summary: 
This PR introduces hyperparameter tuning capabilities for the CMAP project using Weights & Biases (W&B) sweeps. The changes allow users to efficiently explore various model configurations and optimize training processes directly through W&B and SLURM-based workflows.

Changes Introduced:

1.  W&B Sweep Integration:
- Added functionality to define and run W&B sweeps using the sweep_config.yml file located in the configs directory.

- Integrated W&B sweep logic in the training pipeline, supporting distributed sweep agents with SLURM.

2. New Scripts and Configurations:
- configs/sweep_config.yml: Configuration for W&B sweep specifying parameters to tune (e.g., learning rate, batch size, augmentation settings).

- sweep.job: SLURM job script to launch W&B agents for conducting sweeps across multiple trials.

3. Enhanced Training Script:

- Updated train.py to support W&B sweeps, including run_trials() and one_trial() functions.

- Added logging for key metrics like average_test_jaccard_index to ensure they are visible in the W&B dashboard.

- Added argument parsing for hyperparameter tuning and debugging modes.

4. Documentation Updates:

- Expanded the README file to include instructions for running W&B sweeps with SLURM.

- Provided guidance on running sweeps both locally and via SLURM in a cluster environment.

- Included examples for initializing a sweep and starting agents with specific trial counts.

File: README.md [hunk 1]
```diff
@@ -89,6 +89,61 @@ cd /home/YOUR_USERNAME/CMAP
 
 python train.py configs.config --experiment_name <ExperimentName> --aug_type <aug> --split <split> --num_trial <num_trial>
 ```
+### Running Hyperparameter Sweep with SLURM and Weights & Biases
```
Comment (annawoodard) [inline README.md:92]: Please move this closer to the bottom of the README and add another markdown file you link to for "future work" (great idea of yours during today's meeting).

In the future work section, describe what the next steps are to get this working with Slurm.

File: sweep.job [hunk 2]
```diff
@@ -1,19 +1,22 @@
 #!/bin/bash
 #
-#SBATCH --mail-user=USERNAME@cs.uchicago.edu
+#SBATCH --mail-user=xiaoyue1@cs.uchicago.edu
```
Comment (annawoodard) [inline sweep.job:3]: need to remove hardcoded names

File: README.md [hunk 3]
```diff
@@ -89,6 +89,61 @@ cd /home/YOUR_USERNAME/CMAP
...
+wandb agent dsi-clinic-cmap/CMAP/6tmqjcu4 --count 5
+```
+In this example:
+dsi-clinic-cmap is the organization name in W&B.
+CMAP is the project name.
+6tmqjcu4 is the sweep ID.
+--count 5 specifies that the agent should run 5 trials.
+This is particularly useful for quick tests or for running sweeps locally without using a job scheduling system like SLURM.
+
+#### Running a Sweep using a SLURM Job Script
```
Comment (annawoodard) [inline README.md:138]: Move this to the 'future work' section since it doesn't work yet

File: README.md [hunk 4]
```diff
@@ -89,6 +89,61 @@ cd /home/YOUR_USERNAME/CMAP
...
+1. **Edit the sweep.job file** : 
+   - Replace <YOUR-USERNAME> with your username
+   - Replace <W&B_organization_name> and <sweep_id> with the sweep ID obtained from W&B. 
+2. **Submit the sweep.job and Monitor the Job Status**: 
+```
+sbatch sweep.job
+squeue -u <YOUR-USERNAME>
+```
```
Comment (annawoodard) [inline README.md:146]: This readme is missing info about how to actually see your sweep in the wandb interface

Also explain about the confusing plots because we have multiple runs per point

**Review by annawoodard — CHANGES_REQUESTED** [review]
Looks great, please address minor comments and let's get this merged

File: sweep.job (hunk 2 above)
Comment (Wxy-23) [reply sweep.job]: Resolved

File: README.md (hunk 3 above)
Comment (Wxy-23) [reply README.md]: Resolved

File: README.md (hunk 4 above)
Comment (Wxy-23) [reply README.md]: Resolved. Added note to README how to get to W&B interface when running sweep in the terminal. The explanation of the plots are included in the "Monitoring and Interpreting Results" section of future-work.md.

**Comment by Wxy-23** [comment]
Summary
This pull request introduces hyperparameter tuning capabilities for the CMAP project using Weights & Biases (W&B) sweeps and provides detailed documentation in future-work.md. The changes enable efficient exploration of model configurations, enhanced training workflows, and improved guidance for developers.

Changes Introduced
W&B Sweep Integration:

Added functionality to define and run W&B sweeps using configs/sweep_config.yml for hyperparameter tuning.
Integrated distributed sweep logic into the training pipeline with SLURM job support (sweep.job).
Training Pipeline Enhancements:

Updated train.py to include sweep-related functions (run_trials() and one_trial()).
Added logging for metrics like average_test_jaccard_index to improve monitoring in W&B dashboards.
Improved argument parsing for flexible hyperparameter tuning and debugging.
Documentation Updates:

Expanded the README with instructions for running W&B sweeps both locally and via SLURM.
Added examples for initializing sweeps and starting agents with specific trial counts.
Created future-work.md with:
Instructions for resolving SLURM sweep errors.
Guidelines for experiment initialization and monitoring.
Insights into interpreting metrics using parallel coordinate plots.
Updated Files:

configs/sweep_config.yml: Specifies parameters to tune (e.g., learning rate, batch size, augmentation).
sweep.job: SLURM script for running W&B sweep agents.

**Comment by Wxy-23** [comment]
**Summary**
This pull request introduces updates to hyperparameter tuning capabilities for the CMAP project using Weights & Biases (W&B) sweeps and includes detailed documentation in future-work.md. These changes enhance the ability to explore model configurations, streamline training workflows, and provide developers with clear guidance.

**Changes Introduced**
W&B Sweep Integration:
- Updated functionality to define and run W&B sweeps using configs/sweep_config.yml for hyperparameter tuning.

Training Pipeline Enhancements:
- Updated train.py to refine sweep-related functions (run_trials() and one_trial()).
- Improved logging for metrics such as average_test_jaccard_index for better monitoring in W&B dashboards.
- Enhanced argument parsing for flexibility in tuning and debugging.

Documentation Updates:
- Expanded the README with updated instructions for running W&B sweeps locally and via SLURM.
- Added examples for initializing sweeps and starting agents with specific trial counts.
- Created future-work.md to:
    - Provide guidance on resolving SLURM sweep errors.
    - Outline best practices for initializing experiments and monitoring results.
    - Offer insights into interpreting metrics using parallel coordinate plots.

Updates to Supporting Files:
- configs/sweep_config.yml: Updated parameter tuning configurations (e.g., learning rate, batch size, augmentation).
- sweep.job: Refined SLURM script for running W&B sweep agents.

**Review by annawoodard — APPROVED** [review]
Great work @Wxy-23! Thank you!
